### PR TITLE
feat(2346): Update config and model to support job level parameters

### PR DESCRIPTION
**Pull request by sagar1312**
## Context

Ability to define parameters at job level

## Objective

Schema to support job level parameters

## References

https://github.com/screwdriver-cd/screwdriver/issues/2346

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: config/parameters.js [hunk 1]
```diff
@@ -12,17 +12,19 @@ const SCHEMA_PARAMETERS_OBJECT = Joi.object({
...
-        )
+const SCHEMA_PARAMETERS_BASE = Joi.object().pattern(
+    Joi.any(),
+    Joi.alternatives().try(
+        SCHEMA_PARAMETERS_STRING,
+        Joi.array().items(SCHEMA_PARAMETERS_STRING),
+        SCHEMA_PARAMETERS_OBJECT
     )
-    .default({});
+);
```
Comment (pritamstyz4ever) [inline config/parameters.js:22]: Do we need multiple schemas for default and optional?

Comment (jithine) [reply config/parameters.js:22]: not needed, we should be having only one parameters schema

**Review by jithine — CHANGES_REQUESTED** [review]

File: config/parameters.js (hunk 1 above)
Comment (sagar1312) [reply config/parameters.js]: Per discussion with @jithine, pipeline level parameters need not be changed (i.e. it should be not be made as optional).

In the parameters schema, we will export the base schema without `.default({})` or `.optional()`. In pipeline schema, we could extend the base schema to specify the default value `Parameters.parameters.default({})` and for jobs we could extend it to make it optional `Parameters.parameters.default({})`

**Review by jithine — APPROVED** [review]

**Review by pritamstyz4ever — APPROVED** [review]

File: config/job.js [hunk 2]
```diff
@@ -183,7 +185,8 @@ const SCHEMA_JOB = Joi.object()
         sourcePaths: SCHEMA_SOURCEPATHS,
         steps: SCHEMA_STEPS,
         template: SCHEMA_TEMPLATE,
-        templateId: SCHEMA_TEMPLATEID
+        templateId: SCHEMA_TEMPLATEID,
+        parameters: SCHEMA_JOB_PARAMETERS
```
Comment (tkyi) [inline config/job.js:189]: can you put this in alphabetical order

File: config/job.js [hunk 3]
```diff
@@ -223,5 +226,6 @@ module.exports = {
     template: SCHEMA_TEMPLATE,
     templateId: SCHEMA_TEMPLATEID,
     templateJob: SCHEMA_TEMPLATE_JOB,
-    trigger: SCHEMA_TRIGGER
+    trigger: SCHEMA_TRIGGER,
+    parameters: SCHEMA_JOB_PARAMETERS
```
Comment (tkyi) [inline config/job.js:230]: can you put this in alphabetical order

File: api/validator.js [hunk 4]
```diff
@@ -33,6 +33,7 @@ const SCHEMA_JOB_PERMUTATION = Joi.object()
         freezeWindows: Job.freezeWindows,
         image: Job.image,
         order: Job.order,
+        parameters: Parameters.parameters.optional(),
```
Comment (tkyi) [inline api/validator.js:36]:
```suggestion
        parameters: Job.parameters,
```

**Review by tkyi — CHANGES_REQUESTED** [review]
small comments

**Comment by sagar1312** [comment]
@tkyi (@pritamstyz4ever, @jithine )
Addressed the suggestions. Please take a look.

File: api/validator.js (hunk 4 above)
Comment (sagar1312) [reply api/validator.js]: Done

File: config/job.js (hunk 2 above)
Comment (sagar1312) [reply config/job.js]: Done

File: config/job.js (hunk 3 above)
Comment (sagar1312) [reply config/job.js]: Done

**Review by tkyi — APPROVED** [review]

**Pull request merged**